pkg/nodeadm/nodeadm.go

```diff
@@ -35,89 +34,61 @@ type VIPConfiguration struct {
-}
-
 func InitConfigurationForMachine(cluster clusterv1.Cluster, machine clusterv1.Machine, pm spv1.ProvisionedMachine) (*InitConfiguration, error) {
-	cfg := &InitConfiguration{}
+	cfg := &InitConfiguration{
+		MasterConfiguration: KubeadmMasterConfiguration{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "MasterConfiguration",
+				APIVersion: "kubeadm.k8s.io/v1alpha2",
+			},
+		},
+	}
 
 	cpc, err := controller.GetClusterSpec(cluster)
 	if err != nil {
 		return nil, fmt.Errorf("unable to decode cluster spec: %v", err)
 	}
 
-	// MasterConfiguration
-	if cpc.VIPConfiguration.IP != "" {
-		// If ControlPlaneEndpoint is not specified, kubeadm will use AdvertiseAddress + BindPort.
-		cfg.MasterConfiguration.API.ControlPlaneEndpoint = cpc.VIPConfiguration.IP
-		cfg.MasterConfiguration.APIServerCertSANs = []string{cpc.VIPConfiguration.IP}
-	} // else: kubeadm will set defaults
-	cfg.MasterConfiguration.KubernetesVersion = machine.Spec.Versions.ControlPlane
-	cfg.MasterConfiguration.Etcd.Endpoints = []string{"https://127.0.0.1:2379"}
-	cfg.MasterConfiguration.Etcd.CAFile = "/etc/etcd/pki/ca.crt"
-	cfg.MasterConfiguration.Etcd.CertFile = "/etc/etcd/pki/apiserver-etcd-client.crt"
-	cfg.MasterConfiguration.Etcd.KeyFile = "/etc/etcd/pki/apiserver-etcd-client.key"
-	if cpc.ClusterConfig != nil {
-		setInitConfigFromClusterConfig(cfg, cpc.ClusterConfig)
+	mpc, err := controller.GetMachineSpec(machine)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode machine spec: %v", err)
 	}
-	// Networking
-	switch len(cluster.Spec.ClusterNetwork.Pods.CIDRBlocks) {
-	case 0:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at least one block", cluster.Name)
-	case 1:
-		cfg.Networking.PodSubnet = cluster.Spec.ClusterNetwork.Pods.CIDRBlocks[0]
-	case 2:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at most one block", cluster.Name)
+
+	// NodeRegistrationOptions
+	cfg.MasterConfiguration.NodeRegistration.Name = machine.Name
+	cfg.MasterConfiguration.NodeRegistration.Taints = machine.Spec.Taints
+
+	// MasterConfiguration
+	cfg.MasterConfiguration.KubernetesVersion = mpc.ComponentVersions.KubernetesVersion
+	cfg.MasterConfiguration.Etcd.External = &ExternalEtcd{
+		Endpoints: []string{"https://127.0.0.1:2379"},
+		CAFile:    "/etc/etcd/pki/ca.crt",
+		CertFile:  "/etc/etcd/pki/apiserver-etcd-client.crt",
+		KeyFile:   "/etc/etcd/pki/apiserver-etcd-client.key",
 	}
-	switch len(cluster.Spec.ClusterNetwork.Services.CIDRBlocks) {
-	case 0:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at least one block", cluster.Name)
-	case 1:
-		cfg.Networking.ServiceSubnet = cluster.Spec.ClusterNetwork.Services.CIDRBlocks[0]
-	case 2:
-		return nil, fmt.Errorf("cluster %q spec.clusterNetwork.pods.cidrBlocks must contain at most one block", cluster.Name)
+	if err := validateClusterNetworkingConfiguration(cluster); err != nil {
```

Comment:
Since these would be passed through MasterConfiguration and processed by kubeadm, should nodeadm be doing this verification, or can it be offloaded to kubeadm as well?

Reply:
This validates Cluster configuration, which ssh provider must transform to kubeadm configuration.

This is exactly [what the AWS provider does](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/870d03d2d5a25a407b335d3f0b702227dfd7cd59/pkg/cloud/aws/services/kubeadm/aws_defaults.go#L85), though it doesn't validate the Cluster configuration, yet.